src/include/network/connection_handle.h

```diff
@@ -0,0 +1,200 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Terrier
+//
+// connection_handle.h
+//
+// Identification: src/include/network/connection_handle.h
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <arpa/inet.h>
+#include <netinet/tcp.h>
+#include <sys/file.h>
+
+#include <event2/buffer.h>
+#include <event2/bufferevent.h>
+#include <event2/event.h>
+#include <event2/listener.h>
+#include <unordered_map>
+
+#include <csignal>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "common/exception.h"
+#include "loggers/main_logger.h"
+
+#include "network/connection_handler_task.h"
+#include "network/marshal.h"
+#include "network/network_io_wrappers.h"
+#include "network/network_types.h"
+#include "network/postgres_protocol_interpreter.h"
+#include "network/protocol_interpreter.h"
+
+namespace terrier {
+namespace network {
+
+/**
+ * A ConnectionHandle encapsulates all information we need to do IO about
+ * a client connection for its entire duration. This includes a state machine
+ * and the necessary libevent infrastructure for a handler to work on this
+ * connection.
+ */
+class ConnectionHandle {
+ public:
+  /**
+   * Constructs a new ConnectionHandle
+   * @param sock_fd Client's connection fd
+   * @param handler The handler responsible for this handle
+   */
+  ConnectionHandle(int sock_fd, ConnectionHandlerTask *handler);
+
+  DISALLOW_COPY_AND_MOVE(ConnectionHandle);
+
+  /**
+   * @brief Signal to libevent that this ConnectionHandle is ready to handle
+   * events
+   *
+   * This method needs to be called separately after initialization for the
+   * connection handle to do anything. The reason why this is not performed in
+   * the constructor is because it publishes pointers to this object. While the
+   * object should be fully initialized at that point, it's never a bad idea
+   * to be careful.
+   */
+  inline void RegisterToReceiveEvents() {
+    workpool_event_ = conn_handler_->RegisterManualEvent(METHOD_AS_CALLBACK(ConnectionHandle, HandleEvent), this);
+
+    network_event_ = conn_handler_->RegisterEvent(io_wrapper_->GetSocketFd(), EV_READ | EV_PERSIST,
+                                                  METHOD_AS_CALLBACK(ConnectionHandle, HandleEvent), this);
+  }
+
+  /**
+   * Handles a libevent event. This simply delegates the the state machine.
+   */
+  inline void HandleEvent(int, short) { state_machine_.Accept(Transition::WAKEUP, *this); }  // NOLINT
+
+  /* State Machine Actions */
+  // TODO(Tianyu): Write some documentation when feeling like it
+  inline Transition TryRead() { return io_wrapper_->FillReadBuffer(); }
```

Comment:
As a general rule in the new code base `inline` is a bad word. Newer compiler versions pretty much either ignore it or treat it differently from its old semantics. Just remove all of them.